src/parse.jl

```diff
@@ -101,3 +102,523 @@ function ZonedDateTime(str::AbstractString, format::AbstractString; locale::Abst
+            name_end = i
+            state = :simple
+        elseif state == :expanded && c == '>'
+            i = ii
+            state = :closed
+            break
+        else
+            break
+        end
+        i = ii
+    end
+
+    if state == :expanded
+        return ParseNextError("Expected expanded time zone abbreviation end with the greater-than sign (>)", str, prevind(str, name_start), i)
+    elseif state != :closed && name_start > name_end
+        return ParseNextError("Time zone abbreviation must start with a letter or the less-than (<) character", str, i)
+    elseif length(str, name_start, name_end) < 3
+        char_type = if state == :simple
+            "alphabetic characters"
+        else
+            "characters which are either alphanumeric, the plus sign (+), or the minus sign (-)"
+        end
+
+        return ParseNextError("Time zone abbreviation must be at least three $char_type", str, name_start, name_end)
+    else
+        name = SubString(str, name_start, name_end)
+        return name, i
+    end
+end
+
+
+"""
+    _parsesub_offset(str, [i, len]) -> Union{Tuple{Integer, Integer}, Exception}
+
+Parses a section of a string as an offset of the form `[+|-]hh[:mm[:ss]]`. The hour must be
+between 0 and 24, and the minutes and seconds 00 and 59. This follows specification for
+offsets as defined in the tzset(3) man page.
+
+# Example
+```jldoctest
+julia> _parsesub_offset("1:0:0")
+(3600, 5)
+
+julia> _parsesub_offset("-0:1:2")
+(-62, 6)
+```
+"""
+function _parsesub_offset(
+    str::AbstractString,
+    i::Integer=firstindex(str),
+    len::Integer=lastindex(str);
+    name::AbstractString="offset",
+)
+    coefficient = 1
+    hour = minute = second = 0
+
+    if i > len
+        return ParseNextError("Expected $name and instead found end of string", str, i)
+    end
+
+    # Optional sign
+    c, ii = iterate(str, i)::Tuple{Char, Int}
+    if c in ('+', '-')
+        coefficient = c == '-' ? -1 : 1
+        if ii > len
+            return ParseNextError("$(uppercasefirst(name)) sign ($c) is not followed by a value", str, i)
+        end
+        i = ii
+    end
+
+    # Hours
+    val = tryparsenext_base10(str, i, len)
+    if val === nothing
+        return ParseNextError("Expected $name hour digits", str, i)
+    end
+    hour, ii = val
+    if hour < 0 || hour > 24
+        return ParseNextError("Hours outside of expected range [0, 24]", str, i, prevind(str, ii))
+    end
+    i = ii
+    i > len && @goto done
+
+    c, ii = iterate(str, i)::Tuple{Char, Int}
+    c != ':' && @goto done
+    i = ii
+
+    # Minutes
+    val = tryparsenext_base10(str, i, len)
+    if val === nothing
+        return ParseNextError("Expected $name minute digits after colon delimiter", str, i)
+    end
+    minute, ii = val
+    if minute < 0 || minute > 59
+        return ParseNextError("Minutes outside of expected range [0, 59]", str, i, prevind(str, ii))
+    end
+    i = ii
+    i > len && @goto done
```

Comment:
Not sure how others feel about `goto`s, but an alternative could be do make `done` a nested function, and just call that here while passing along `i`?

Reply:
The `goto`s are mainly being used for performance reasons. Even if I make a nested function I'd still need to do add a `return` which I'm not sure would help with clairity

Reply:
Hmm I don't have permissions to resolve my own comments, but I'm okay with keeping them in if this is the case.